litex/soc/software/include/hw/common.h

```diff
@@ -14,37 +16,255 @@
 }
 
-static inline void csr_writel(uint32_t value, unsigned long addr)
+static inline uint16_t csr_rd_uint16(unsigned long a)
 {
-	*((volatile uint32_t *)addr) = value;
+	return _csr_rd((unsigned long *)a, sizeof(uint16_t));
 }
 
-static inline uint32_t csr_readl(unsigned long addr)
+static inline uint32_t csr_rd_uint32(unsigned long a)
 {
-	return *(volatile uint32_t *)addr;
+	return _csr_rd((unsigned long *)a, sizeof(uint32_t));
+}
+
+static inline uint64_t csr_rd_uint64(unsigned long a)
+{
+	return _csr_rd((unsigned long *)a, sizeof(uint64_t));
+}
+
+static inline void csr_wr_uint8(uint8_t v, unsigned long a)
+{
+	_csr_wr((unsigned long *)a, v, sizeof(uint8_t));
+}
+
+static inline void csr_wr_uint16(uint16_t v, unsigned long a)
+{
+	_csr_wr((unsigned long *)a, v, sizeof(uint16_t));
+}
+
+static inline void csr_wr_uint32(uint32_t v, unsigned long a)
+{
+	_csr_wr((unsigned long *)a, v, sizeof(uint32_t));
+}
+
+static inline void csr_wr_uint64(uint64_t v, unsigned long a)
+{
+	_csr_wr((unsigned long *)a, v, sizeof(uint64_t));
+}
+
+/* byteswap methods for each distinct unsigned integer (uintX_t) C type */
+#define _bswp_uint8_t(x) (x)
+#define _bswp_uint16_t(x) bswap_16(x)
+#define _bswp_uint32_t(x) bswap_32(x)
+#define _bswp_uint64_t(x) bswap_64(x)
+
+/* variable-width (dynamic) byteswap (actual width only known at run time) */
+static inline uint64_t _bswp_dyn_X(uint64_t n, int x)
+{
+	switch (x) {
+	case 2:
+		return bswap_16(n);
+	case 4:
+		return bswap_32(n);
+	case 8:
+		return bswap_64(n);
+	}
+	return n;
+}
+
+/* Read a CSR located at address 'a' into an array 'buf' of 'cnt' elements
+ * of type 'uintX_t'.
+ *
+ * NOTE: Since CSR_DW_BYTES is a constant here, we might be tempted to further
+ * optimize things by leaving out one or the other of the if() branches below,
+ * depending on each unsigned type width;
+ * However, this code is also meant to serve as a reference for how CSRs are
+ * to be manipulated by other programs (e.g., an OS kernel), which may benefit
+ * from dynamically handling multiple possible CSR subregister data widths
+ * (e.g., by passing a value in through the Device Tree).
+ * Ultimately, if CSR_DW_BYTES is indeed a constant, the compiler should be
+ * able to determine on its own whether it can automatically optimize away one
+ * of the if() branches! */
+#define _csr_rd_buf(a, uintX_t, buf, cnt) \
+{ \
+	int i, j, nsubs, n_sub_elem; \
+	unsigned long *addr = (unsigned long *)(a); \
+	uint64_t r; \
+	if (sizeof(uintX_t) >= CSR_DW_BYTES) { \
+		/* one or more subregisters per element */ \
+		for (i = 0; i < cnt; i++) { \
+			buf[i] = _csr_rd(addr, sizeof(uintX_t)); \
+			addr += num_subregs(sizeof(uintX_t)); \
+		} \
+	} else { \
+		/* multiple elements per subregister (2, 4, or 8) */ \
+		nsubs = num_subregs(sizeof(uintX_t) * cnt); \
+		n_sub_elem = CSR_DW_BYTES / sizeof(uintX_t); \
+		for (i = 0; i < nsubs; i++) { \
+			r = _bswp_dyn_X(addr[i], CSR_DW_BYTES); \
+			for (j = 0; j < n_sub_elem; j++) { \
```

Comment:
I think if I go backwards `for (j = n_sub_elem - 1; j >= 0; j--)` I might be able to forego having to use byteswap routines at all, will try that and force-push an updated patch if successful...